packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -36,18 +37,86 @@ enum _ScaffoldSlot {
   statusBar,
 }
 
+/// Geometry information for scaffold components.
+///
+/// To get a [ValueNotifier] for the scaffold geometry call
+/// [Scaffold.geometryOf].
+@immutable
+class ScaffoldGeometry {
+  const ScaffoldGeometry({
+    this.bottomNavigationBarTop,
+    this.floatingActionButtonPosition,
+    this.floatingActionButtonScale = 1.0,
+  });
+
+  /// The distance from the scaffold's top edge to the top edge of the
+  /// rectangle in which the [Scaffold.bottomNavigationBar] bar is being laid
+  /// out.
+  ///
+  /// When there is no [Scaffold.bottomNavigationBar] set, this will be null.
+  final double bottomNavigationBarTop;
+
+  /// The rectangle in which the scaffold is laying out
+  /// [Scaffold.floatingActionButton].
+  ///
+  /// The floating action button might be scaled inside this rectangle, to get
+  /// the bounding rectangle in which the floating action is painted scale this
+  /// value by [floatingActionButtonScale].
+  ///
+  /// This can have any value when there is no [Scaffold.floatingActionButton]
+  /// set.
+  final Rect floatingActionButtonPosition;
+
+  /// The amount by which the [Scaffold.floatingActionButton] is scaled.
+  ///
+  /// To et the bounding rectangle in shich the floating action button is
+  /// painted scale [floatingActionPosition] by this proportion. 
+  final double floatingActionButtonScale;
+
+  static ScaffoldGeometry _copyWith({
+    ScaffoldGeometry previousGeometry,
+    double bottomNavigationBarTop,
+    Rect floatingActionButtonPosition,
+    double floatingActionButtonScale,
+  }) {
+    return new ScaffoldGeometry(
+      bottomNavigationBarTop: bottomNavigationBarTop ??  previousGeometry?.bottomNavigationBarTop,
+      floatingActionButtonPosition: floatingActionButtonPosition ??  previousGeometry?.floatingActionButtonPosition,
+      floatingActionButtonScale: floatingActionButtonScale ??  previousGeometry?.floatingActionButtonScale,
+    );
+  }
+}
+
+class _ScaffoldGeometryNotifier extends ValueNotifier<ScaffoldGeometry> {
+  _ScaffoldGeometryNotifier(ScaffoldGeometry geometry)
+    : super(geometry);
+
+  @override
+  ScaffoldGeometry get value{
+    if (!RendererBinding.instance.pipelineOwner.debugDoingPaint)
+      throw new FlutterError(
+        'Scaffold.geometryOf() value was accessed not during the paint phase.'
```

Comment:
we should phrase this as something like "Scaffold.geometryOf() value must only be accessed during the paint phase.", otherwise the logical reaction to that error message is "I know, so what".

Reply:
Maybe even make this a multiline message:

```
Scaffold.geometryOf() value obtained outside paint phase.
The ScaffoldGeometry is only available during the paint phase, because its value is computed during
the animation and layout phases prior to painting. If you are trying to obtain the value during the
build or layout phases, consider moving your logic to the paint phase, since the value is not yet
completely available during build or layout. If you are trying to obtain the value between frames,
consider caching the value during paint.
```

...or something

Reply:
I sub-sampled parts of your message :)